---
title: "COVID-19 GWAS meta-analysis and eQTLs for CRISPR screen hits - freeze 5 data"
author: "Silva Kasela"
date: "05/12/2021"
output: html_document
---

```{r args, echo=FALSE}
# n top-ranked genes to include
n_rank <- commandArgs(trailingOnly = TRUE)
n_rank <- as.numeric(n_rank)

```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.path = here::here("covid_gwas", "freeze5", "enrichment", paste0("fig_ntop_", n_rank, "/")), dev = c('png', 'pdf'))
options(knitr.table.format = "html")

library(here)
library(data.table)
library(ggplot2)
library(ggrastr)
library(knitr)
library(kableExtra)

source(here("covid_gwas", "qqplot_confbands.R"))

# Theme for ggplot2
theme_set(theme_classic() +
          theme(plot.title = element_text(size = 14),
	              plot.subtitle = element_text(size = 12),
                axis.title = element_text(size = 12),
                axis.title.y = element_text(vjust = 2),
                axis.title.x = element_text(vjust = -0.75),
                axis.text = element_text(size = 11),
                strip.text = element_text(size = 11),
                legend.title = element_text(size = 12),
                legend.text = element_text(size = 11)))

# Def functions
lambda_quantile <- function(p, q = 0.5) quantile(qchisq(p, df = 1, lower.tail = FALSE), probs = q) / qchisq(q, df = 1)

```

Question: do we see inflation in COVID-19 GWAS data for the GTEx lung eVariants for CRISPR screen hit genes?

**QQ plot for**

1. all variants in the GWAS (full background),
2. all lead lung eVariants (eQTL background),
3. eVariants of CRISPR screen hit genes

**Quantifying enrichment**

Using lambda to quantify enrichment of eQTLs in GWAS

* _Lambda_ – enrichment estimate, inflation of test statistics
    + lambda_0.5 – relative to the chi-square quantile function of 0.5 (median – classical inflation marker)
    + lambda_0.1 – relative to the chi-square quantile function of 0.1 (10th percentile – 10% of the most significant tests)

<!-- Lambda_0.5 is defined as the median of the resulting chi-squared test statistics divided by the expected median of the chi-squared distribution. The median of a chi-squared distribution with one degree of freedom is 0.4549364.     -->

**Significance of the observed enrichment**

Using permutation based test to estimate the significance of the enrichment.

* Select _n_ eVariants from the eQTL Lung All set _k_ = 10,000 times (including the original set of CRISPR eQTLs), where _n_ is the number of variants that are CRISPR eQTLs and used in the COVID-19 GWAS
* Calculate permutation P-value as the proportion of permuted enrichment as extreme as or more extreme than the true (observed) enrichment
    + Note: taking log() of permuted lambda values to get a more symmetrical null distribution around 0 to calculate two-sided permutation P-values

## Data

#### Top-ranked genes from the CRISPR screen that have eQTLs in GTEx Lung

```{r}
# Gencode gene annotations
gene_annot <- read.table("~/gtex_v8/references/gencode.v26.GRCh38.genes.gtf", header = F, sep = "\t", stringsAsFactors = F)
gene_annot <- gene_annot[gene_annot$V3 %in% "gene",]
gene_annot$gene_id <- sapply(gene_annot$V9, function(x) sub("gene_id ", "", unlist(strsplit(x, "; "))[1]))
gene_annot$gene_type <- sapply(gene_annot$V9, function(x) sub("gene_type ", "", unlist(strsplit(x, "; "))[3]))

```

__All lead Lung eVariants (as a background)__

* Keep only protein-coding genes

```{r}
gtex_lung <- fread("~/gtex_v8/eqtl/GTEx_Analysis_v8_eQTL/Lung.v8.egenes.txt.gz", header = T, sep = "\t", stringsAsFactors = F, data.table = F)
gtex_lung <- gtex_lung[gtex_lung$qval < 0.05,]
# X to 23
gtex_lung$chr <- ifelse(gtex_lung$chr %in% "chrX", 23, sub("chr", "", gtex_lung$chr))
# SNP ID - chr:pos:ref:alt
gtex_lung$id <- with(gtex_lung, paste(chr, variant_pos, ref, alt, sep = ":"))
```

* `r nrow(gtex_lung)` significant eGenes in Lung (FDR < 0.05)

```{r}
# Keep only protein-coding genes
stopifnot(gtex_lung$gene_id %in% gene_annot$gene_id)
gtex_lung$gene_type <- gene_annot[match(gtex_lung$gene_id, gene_annot$gene_id), "gene_type"]
# print(addmargins(table(gtex_lung$gene_type)))
gtex_lung <- gtex_lung[gtex_lung$gene_type %in% "protein_coding", ]

```

* `r nrow(gtex_lung)` significant protein-coding eGenes in Lung (FDR < 0.05)


__Top-ranked genes__ from the CRISPR screen by [Daniloski et al., 2021](https://doi.org/10.1016/j.cell.2020.10.030)

  * Considering `r n_rank` top-ranked genes by each of the three methods: RRA, RIGER, SBR

```{r}
crispr_genes <- read.table(here("data", "top_500_genes_SB_RIGER_RRA_with_ensembl_id.intersect_gtex_lung.txt"), header = T, sep = "\t", stringsAsFactors = F)
crispr_genes <- crispr_genes[crispr_genes$is_egene_in_gtex_lung %in% 1, ] # 439
stopifnot(crispr_genes$gene_id %in% gtex_lung$gene_id)
crispr_genes$id <- gtex_lung[match(crispr_genes$gene_id, gtex_lung$gene_id), "id"]

```

* `r nrow(crispr_genes)` of the top-ranked CRISPR genes have eQTLs in Lung

```{r crispr-genes-filter}
if (n_rank < 500) {
  sel <- crispr_genes[, 3:5]
  if (!class(sel) %in% "matrix") {
    sel <- as.matrix(sel)
  }
  rownames(sel) <- crispr_genes$gene_id

  keep_genes <- apply(sel, 1, function(x) {
    sum(x <= n_rank, na.rm = T)
  })
  keep_genes <- names(keep_genes[keep_genes > 0])
  crispr_genes <- crispr_genes[crispr_genes$gene_id %in% keep_genes,]
}

```

* `r nrow(crispr_genes)` of the top-ranked CRISPR hit genes with eQTLs in Lung are among top `r n_rank`


#### COVID-19 host genetics initiative

See the website for more info [here](https://www.covid19hg.org/).

> The COVID-19 host genetics initiative brings together the human genetics community to generate, share and analyze data to learn the genetic determinants of COVID-19 susceptibility, severity and outcomes.

**COVID19-hg GWAS meta-analyses round 5 (released on January 18, 2021):**

> Meta-analysis was done with fixed effects inverse variance weighting. Results are available in genome builds 38 and 37. An AF filter of 0.001 and an INFO filter of 0.6 was applied to each study before meta.

Meta-analysis results availble for the following phenotypes ([ref](https://docs.google.com/document/d/1okamrqYmJfa35ClLvCt_vEe4PkvrTwggHq7T3jbeyCI/edit)):

* __Analysis A: VERY SEVERE RESPIRATORY CONFIRMED COVID__
    + __Analysis A2_v2 (name ANA_A2_V2)__
        * Cases: Hospitalized laboratory confirmed SARS-CoV-2 infection (RNA and/or serology based), AND (death OR respiratory support (intubation, CPAP, BiPAP, CNP (continue external negative pressure), Optiflow/very high flow Positive End Expiratory Pressure Oxygen - (see Appendix 1 for suggestive codes) AND hospitalization with COVID19 as primary reason for admission.
        * Controls: everybody that is not a case, e.g. population

* __Analysis B: HOSPITALIZED LAB CONFIRMED COVID__
    + __Analysis B1_v2 (name ANA_B1_V2)__
        * Cases: Hospitalized laboratory confirmed SARS-CoV-2 infection (RNA and/or serology based), hospitalization due to corona-related symptoms.
        * Controls: Laboratory confirmed SARS-CoV-2 infection (RNA and/or serology based) AND not hospitalised 21 days after the test.
    + __Analysis B2_v2 (name ANA_B2_V2)__
        * Cases: As B1_v2
        * Controls: everybody that is not a case, e.g. population

* __Analysis C: PARTIAL-SUSCEPTIBILITY__
    + __Analysis C2_v2 (name ANA_C2_V2)__
        * Cases: individuals with laboratory confirmation of SARS-CoV-2 infection (RNA and/or serology based) OR EHR/ICD coding/ Physician Confirmed COVID-19 (See Appendix 1 for suggestive codes) OR self-reported COVID-19 positive (e.g. by questionnaire)
        * Controls: everybody that is not a case, e.g. population

Analysis model ([ref](https://docs.google.com/document/d/1Pcq1jttF8W7ifEUXA6-a1WVMsUyEoAybS6IqvuP-Uv8/edit#heading=h.lv1f4316vktu)):

* Phenotype ~ variant + age + age2 + sex + age*sex + PCs + study_specific_covariates

```{r}
files_gwas <- c("https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_A2_ALL_leave_23andme_20210107.txt.gz",
                "https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_A2_ALL_eur_leave_23andme_20210107.txt.gz",
                "https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_B1_ALL_leave_23andme_20210107.txt.gz",
                "https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_B1_ALL_eur_leave_23andme_20210107.txt.gz",
                "https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_B2_ALL_leave_23andme_20210107.txt.gz",
                "https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_B2_ALL_eur_leave_23andme_20210107.txt.gz",
                "https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_C2_ALL_leave_23andme_20210107.txt.gz",
                "https://storage.googleapis.com/covid19-hg-public/20201215/results/20210107/COVID19_HGI_C2_ALL_eur_leave_23andme_20210107.txt.gz")

gwas_df <- data.frame(code = sapply(files_gwas, function(x) unlist(strsplit(x, "_"))[3]),
                      population = sapply(files_gwas, function(x) paste(unlist(strsplit(sub("_leave_23andme_20210107.txt.gz", "", x), "_"))[-c(1:3)], collapse = "_")),
                      name = "NA",
                      file_name = sapply(files_gwas, function(x) unlist(strsplit(x, "/"))[8]),
                      file_path = files_gwas,
                      stringsAsFactors = F)
rownames(gwas_df) <- 1:nrow(gwas_df)
gwas_df$name <-  ifelse(gwas_df$code %in% "A1", "very severe respiratory confirmed covid vs. not hospitalized covid",
                        ifelse(gwas_df$code %in% "A2", "very severe respiratory confirmed covid vs. population",
                               ifelse(gwas_df$code %in% "B1", "hospitalized covid vs. not hospitalized covid",
                                      ifelse(gwas_df$code %in% "B2", "hospitalized covid vs. population",
                                             ifelse(gwas_df$code %in% "C1", "covid vs. lab/self-reported negative",
                                                    ifelse(gwas_df$code %in% "C2", "covid vs. population",
                                                           ifelse(gwas_df$code %in% "D1", "predicted covid from self-reported symptoms vs. predicted or self-reported non-covid", NA)))))))

kable(gwas_df) %>%
  kable_styling(bootstrap_options = c("striped", "condensed"), full_width = F, font_size = 14)

```

### Enrichment of COVID-19 GWAS signal

```{r fig-covid19-gwas-enrichment, fig.width=4.5, fig.height=4.5, out.width="33%", results='asis', warning=FALSE}
set.seed(1205)
# Number of permutations + 1 (original set of CRISPR eQTLs)
k <- 9999
gwas_df$lambda_0.1 <- NA
gwas_df$perm_p_lambda_0.1 <- NA
gwas_df$perm_p_log_lambda_0.1 <- NA
gwas_eqtl_lead <- list()

for (i in 1:nrow(gwas_df)) {
  cat(" \n", fill = T)
  cat("**", gwas_df$code[i], " (", gwas_df$population[i], ") - ", gwas_df$name[i], "** \n", fill = T)
  gwas <- fread(gwas_df$file_path[i], header = T, sep = "\t", stringsAsFactors = F, data.table = F)
  cat(nrow(gwas), "variants in the GWAS file \n", fill = T)

  # all lead lung eVariants (eQTL background)
  sel_all_lead <- gwas[gwas$SNP %in% gtex_lung$id,]
  cat(nrow(sel_all_lead), "out of", nrow(gtex_lung), "lead Lung eVariants found from GWAS \n", fill = T)
  gwas_eqtl_lead[[i]] <- sel_all_lead
  names(gwas_eqtl_lead)[i] <- paste0(gwas_df$code[i], "_", gwas_df$population[i])

  # eVariants of CRISPR screen hit genes
  sel_crispr_hits <- sel_all_lead[sel_all_lead$SNP %in% crispr_genes$id,]
  cat(nrow(sel_crispr_hits), "out of", nrow(crispr_genes), "lead eVariants of CRISPR screen hit genes found from GWAS \n", fill = T)

  # Enrichment statistics - lambda
  lambda_all <- round(as.numeric(lambda_quantile(p = sel_all_lead$all_inv_var_meta_p, q = 0.1)), 2)
  lambda_crispr <- round(as.numeric(lambda_quantile(p = sel_crispr_hits$all_inv_var_meta_p, q = 0.1)), 2)

  cat("Lambda 0.5 - eQTL Lung all - ", round(as.numeric(lambda_quantile(p = sel_all_lead$all_inv_var_meta_p)), 2), " \n", fill = T)
  cat("Lambda 0.5 - eQTL Lung CRISPR hits - ", round(as.numeric(lambda_quantile(p = sel_crispr_hits$all_inv_var_meta_p)), 2), " \n", fill = T)

  # Permutation test
  n <- nrow(sel_crispr_hits)
  perm_lambda_0.1 <- rep(NA, k)

  for (j in 1:k) {
    idx <- sample(1:nrow(sel_all_lead), n, replace = FALSE)
    perm_p <- sel_all_lead[idx, "all_inv_var_meta_p"]
    perm_lambda_0.1[j] <- as.numeric(lambda_quantile(p = perm_p, q = 0.1))
  }

  df_perm <- data.frame(lambda = perm_lambda_0.1,
                        lambda_log = log(perm_lambda_0.1),
                        stringsAsFactors = F)
  perm_p_lambda_0.1 <- (sum(perm_lambda_0.1 >= lambda_crispr) + 1) / (k + 1)
  # For two-sided permutation P-value we need to take log() of lambdas to get a null distribution symmetrical around 0
  perm_p_log_lambda_0.1 <- (sum(abs(log(perm_lambda_0.1)) >= abs(log(lambda_crispr))) + 1) / (k + 1)

  # Results to df
  gwas_df[i, 6:8] <- c(lambda_crispr, perm_p_lambda_0.1, perm_p_log_lambda_0.1)

  # Plot figure
  qq <- qqplot_confbands(pval = gwas$all_inv_var_meta_p,
                        confbands = FALSE,
                        main = paste0(gwas_df$code[i], " ", gwas_df$population[i], " GWAS - eQTLs in Lung"),
                        color_group_name = "GWAS",
                        add_p = list(x = sel_all_lead$all_inv_var_meta_p, y = sel_crispr_hits$all_inv_var_meta_p),
                        add_group = list(paste0("eQTL Lung all, l = ", lambda_all),
                                         paste0("eQTL Lung CRISPR hits, l = ", lambda_crispr)),
                        raster_plot = TRUE,
                        point_size = 3.5)
  # Modify legend
  cols <- c("darkorchid", "forestgreen", "black")
  names(cols) <- c(paste0("eQTL Lung CRISPR hits, l = ", lambda_crispr), paste0("eQTL Lung all, l = ", lambda_all), "GWAS")
  my_labs <- list(bquote("eQTL Lung CRISPR hits, " ~ lambda[0.1] == ~ .(lambda_crispr)),
                 bquote("eQTL Lung all, " ~ lambda[0.1] == ~.(lambda_all)),
                 "GWAS")

  qq <- qq +
   guides(color = guide_legend(title.position = "right")) + # hack to get legend on the topright corner without leaving space for legend title; legend.title = element_blank() still leaves some empty space on the top
   scale_color_manual(name = "",
                      breaks = c(paste0("eQTL Lung CRISPR hits, l = ", lambda_crispr),
                                 paste0("eQTL Lung all, l = ", lambda_all),
                                 "GWAS"),
                      values =  cols,
                      labels = my_labs) +
   theme(legend.justification = c(0, 1),
         legend.position = c(0, 1),
         legend.background = element_rect(fill = NA))
  print(qq)

  # Histogram - All eQTLs
  g_hist_1 <- ggplot(data = sel_all_lead, aes(x = all_inv_var_meta_p)) +
    labs(title = paste0(gwas_df$code[i], " ", gwas_df$population[i], " GWAS - eQTL Lung all\n", nrow(sel_all_lead), " eVariants"),
         x = "GWAS P-value for selected eVariants",
         y = "Count") +
    geom_histogram(col = "forestgreen", fill = "white", breaks = seq(0, 1, by = 1/20))
  print(g_hist_1)

  # Histogram - CRISPR genes with eQTLs
  g_hist_2 <- ggplot(data = sel_crispr_hits, aes(x = all_inv_var_meta_p)) +
    labs(title = paste0(gwas_df$code[i], " ", gwas_df$population[i], " GWAS - eQTL Lung CRISPR hits\n", nrow(sel_crispr_hits), " eVariants"),
         x = "GWAS P-value for selected eVariants",
         y = "Count") +
    geom_histogram(col = "darkorchid", fill = "white", breaks = seq(0, 1, by = 1/20)) +
    scale_x_continuous(breaks = seq(0, 1, 0.2))
  print(g_hist_2)

  # Histogram of permuted lambda_0.1
  label <- bquote(lambda[0.1] == ~ .(lambda_crispr))
  g_perm_hist_1 <- ggplot(data = df_perm, aes(x = lambda)) +
    labs(title = paste0(gwas_df$code[i], " ", gwas_df$population[i], " GWAS - eQTL Lung CRISPR hits\n", nrow(sel_crispr_hits), " eVariants"),
         subtitle = paste0("Permutation P-value = ", signif(perm_p_lambda_0.1, 3)),
         x = bquote("Permuted " ~ lambda[0.1]),
         y = "Count") +
    geom_histogram(col = "darkorchid", fill = "white", bins = 30) +
    geom_vline(xintercept = lambda_crispr, col = "black", lty = 2, lwd = 0.75) +
    annotate(geom = "text", x = lambda_crispr, y = Inf, label = label, vjust = 2, hjust = -0.1)
  print(g_perm_hist_1)

  # Histogram of permuted log(lambda_0.1) - two-sided P-values
  label <- bquote(log(lambda[0.1]) == ~ .(round(log(lambda_crispr), 2)))
  g_log_perm_hist_1 <- ggplot(data = df_perm, aes(x = lambda_log)) +
    labs(title = paste0(gwas_df$code[i], " ", gwas_df$population[i], " GWAS - eQTL Lung CRISPR hits\n", nrow(sel_crispr_hits), " eVariants"),
         subtitle = paste0("Permutation P-value = ", signif(perm_p_log_lambda_0.1, 3)),
         x = bquote("Permuted " ~ log(lambda[0.1])),
         y = "Count") +
    geom_histogram(col = "darkorchid", fill = "white", bins = 30) +
    geom_vline(xintercept = log(lambda_crispr), col = "black", lty = 2, lwd = 0.75) +
    annotate(geom = "text", x = log(lambda_crispr), y = Inf, label = label, vjust = 2, hjust = ifelse(log(lambda_crispr) > 0.1, -0.1, 1.1))
  print(g_log_perm_hist_1)

  # Clean up
  rm(lambda_all, lambda_crispr, perm_lambda_0.1)

}

```

```{r write-out}
write.table(gwas_df, file = here("covid_gwas", "freeze5", "enrichment", paste0("covid19_gwas_analysis_round5.top_", n_rank, "_genes.txt")), col.names = T, row.names = F, sep = "\t", quote = F)
save(gwas_eqtl_lead, file = here("covid_gwas", "freeze5", "enrichment", paste0("gwas_eqtl_lead.top_", n_rank, "_genes.RData")))

```
