---
title: "Summary of eQTL associations on chr3 GWAS locus - B2 analysis"
author: "Silva Kasela"
date: "6/20/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, fig.path = here::here("covid_gwas", "freeze5", "chr3_coloc", "fig_summary_coloc_B2/"), dev = c('png', 'pdf'))
options(knitr.table.format = "html") 

library(here)
library(dplyr)
library(tidyr)
library(ggplot2)
library(pheatmap)
library(knitr)
library(kableExtra)

# ggplot2 theme
theme_set(theme_classic() +
            theme(plot.title = element_text(size = 14),
                  plot.subtitle = element_text(size = 12),
                  axis.title = element_text(size = 12),
                  axis.title.y = element_text(vjust = 2),
                  axis.title.x = element_text(vjust = -0.75),
                  axis.text = element_text(size = 11),
                  strip.text = element_text(size = 10),
                  legend.title = element_text(size = 12),
                  legend.text = element_text(size = 11)))

```

Can we detect colocalization signal between COVID-19 GWAS and eQTLs?

```{r}
gwas_analysis <- "B2_ALL_leave_23andme"
lead_gwas <- "3:45823240:T:C"

```


Data

* COVID-19 HGI (Freeze 5 data): `r gwas_analysis` - hospitalized covid vs. population GWAS summary statistics
* eQTL Catalogue: eQTL summary statistics from various (immune) cell types
* GTEx v8: eQTL summary statistics from 49 tissues

Approach

Run coloc between COVID-19 GWAS and eQTLs, if thereâ€™s an eVariant within 100kb of the GWAS variant with nominal p-value < 10-4

* Coloc-standard
    + Run coloc in a 1Mb-region centered on the lead GWAS variant (+/- 500kb from the lead GWAS variant)
        * Priors used: `p1` = 1e-4, `p2` = 1e-4, `p12` = 5e-6
* Coloc-cond/mask
    + Run coloc in a 2Mb-region centered on the lead GWAS variant (+/- 1Mb from the lead GWAS variant)
    + Conditioning used for GTEx v8 with LD data from GTEx, masking used for eQTL catalogue data with LD from 1000G EUR population
        * `p1` = 1e-4, `p2` = 1e-4, `p12` = 5e-6
        * `mode` = 'iterative', `pthr` = 1e-04 (P-value threshold for secondary signals), `r2thr` = 0.01 (LD threshold for masking), `maxhits` = 3

```{r}
res_list <- list()
res_condmask_list <- list()

```

```{r read-in-data-eqtl-catalogue}
# eQTL Catalogue studies - eQTL Catalogue
tabix_paths <- read.table("https://raw.githubusercontent.com/eQTL-Catalogue/eQTL-Catalogue-resources/master/tabix/tabix_ftp_paths.tsv", sep = "\t", header = TRUE, stringsAsFactors = FALSE)
tabix_paths <- tabix_paths %>%
  filter(quant_method %in% c("ge", "microarray"),
         !study %in% "GTEx")
tabix_paths$id <- sapply(tabix_paths$ftp_path, function(x) unlist(strsplit(x, "/"))[11])
tabix_paths$id <- sapply(tabix_paths$id, function(x) unlist(strsplit(x, "[.]"))[1])
print(addmargins(table("eQTL Catalogue studies" = tabix_paths$quant_method)))

# coloc-standard -------------------------
# Results of coloc
res <- read.table(here("covid_gwas", "freeze5", "chr3_coloc", "result", paste0("summary_coloc.", gwas_analysis, ".eQTL_catalogue.txt")), header = T, sep = "\t", stringsAsFactors = F)
res$id <- res$eqtl_data
res$id_match <- paste0(res$eqtl_data, res$eqtl_gene_name)
stopifnot(tabix_paths$id %in% res$id)
cat("Unique eQTL studies in the eQTL Catalogue - ", length(unique(res$eqtl_data)), fill = TRUE)

# filter to take best probe = probe with min lead_gwas_variant_eqtl_pval
res <- res %>% 
  group_by(eqtl_data, eqtl_gene_name) %>% 
  mutate(min_lead_gwas_variant_eqtl_pval = min(lead_gwas_variant_eqtl_pval)) %>%
  arrange(lead_gwas_variant_eqtl_pval, .by_group = TRUE) %>% 
  ungroup() %>% 
  distinct(eqtl_gene_name, eqtl_data, .keep_all = TRUE) %>% 
  data.frame()

res_list[[1]] <- res
names(res_list)[1] <- "eQTL_catalogue"

# coloc-condmask -------------------------
# Results of coloc
res <- read.table(here("covid_gwas", "freeze5", "chr3_coloc", "result_condmask", paste0("summary_coloc_condmask.", gwas_analysis, ".eQTL_catalogue.txt")), header = T, sep = "\t", stringsAsFactors = F)
res$id <- res$eqtl_data
res$id_match <- paste0(res$eqtl_data, res$eqtl_gene_name)

# filter to take best probe = probe with min lead_gwas_variant_eqtl_pval => not needed
# only one probe per gene in condmask files (i.e., only one probe per gene for coloc test anyway)
id <- paste0(res$eqtl_data, "_", res$eqtl_gene_name, "_", res$hit2)
stopifnot(length(unique(id)) == length(id))
rm(id)

# filter to take best indep signal = hit2 with max pp4
res <- res %>% 
  group_by(eqtl_data, eqtl_gene_name) %>% 
  mutate(max_pp4 = max(PP.H4.abf)) %>%
  # arrange(desc(PP.H4.abf), .by_group = TRUE) %>% 
  ungroup() %>% 
  # distinct(eqtl_gene_name, eqtl_data, .keep_all = TRUE) %>% 
  select(-starts_with("coloc_")) %>% 
  data.frame()

res_condmask_list[[1]] <- res
names(res_condmask_list)[1] <- "eQTL_catalogue"

```

```{r read-in-data-gtex}
# coloc-standard---------------
# Results of coloc - GTEX v8
res <- read.table(here("covid_gwas", "freeze5", "chr3_coloc", "result", paste0("summary_coloc.", gwas_analysis, ".GTEx_v8.txt")), header = T, sep = "\t", stringsAsFactors = F)
res$id <- sapply(res$eqtl_data, function(x) sub("GTEx_v8_ge_", "", x))
res$id_match <- paste0(res$eqtl_data, res$eqtl_gene_name)
cat("Unique eQTL studies in GTEx v8 - ", length(unique(res$eqtl_data)), fill = TRUE)

res_list[[2]] <- res
names(res_list)[2] <- "GTEx_v8"

# coloc-condmask ------------
# Results of coloc - GTEX v8
res <- read.table(here("covid_gwas", "freeze5", "chr3_coloc", "result_condmask", paste0("summary_coloc_condmask.", gwas_analysis, ".GTEx_v8.txt")), header = T, sep = "\t", stringsAsFactors = F)
res$id <- sapply(res$eqtl_data, function(x) sub("GTEx_v8_ge_", "", x))
res$id_match <- paste0(res$eqtl_data, res$eqtl_gene_name)

# filter to take best indep signal = hit2 with max pp4
res <- res %>% 
  group_by(eqtl_data, eqtl_gene_name) %>% 
  mutate(max_pp4 = max(PP.H4.abf)) %>%
  # arrange(desc(PP.H4.abf), .by_group = TRUE) %>% 
  ungroup() %>% 
  # distinct(eqtl_gene_name, eqtl_data, .keep_all = TRUE) %>% 
  select(-starts_with("coloc_")) %>% 
  data.frame()

res_condmask_list[[2]] <- res
names(res_condmask_list)[2] <- "GTEx_v8"

```

## Summary of the results

#### P-value based lookup

What's the eQTL P-value of the lead eVariant and lead GWAS variant (`r lead_gwas`) in various studies from the eQTL Catalogue and GTEx?

* If gene expression data from microarray, then taking the probe for which the GWAS variant has the lowest eQTL P-value

__1) Heatmap of lead eVariant P-value in a 3Mb region centered at the GWAS variant__

* eQTL Catalogue

```{r fig-eqtl-pval-heatmap-eqtl-catalogue, fig.width=12, fig.height=5}
i <- 1
res <- res_list[[i]]
plot_data <- res %>%
    select(eqtl_gene_name, id, lead_eqtl_pval_larger_window) %>%
    mutate(pval = -log10(lead_eqtl_pval_larger_window )) %>%
    pivot_wider(-lead_eqtl_pval_larger_window, names_from = id, values_from = pval) %>%
    tibble::column_to_rownames(var = "eqtl_gene_name") %>% 
    as.matrix()

pheatmap(mat = plot_data[c("LIMD1", "SACM1L", "SLC6A20", "LZTFL1", "CCR9", "FYCO1", "CXCR6", "XCR1"), ],
         scale = "none",
         color = viridis::viridis(100),
         na_col = "grey95",
         cluster_rows = F, cluster_cols = T)

#ggplot(data = res, aes(x = eqtl_gene_name, y = eqtl_data, fill = -log10(lead_eqtl_pval_larger_window))) + 
#  geom_tile()

```

* GTEx

```{r fig-eqtl-pval-heatmap-gtex, fig.width=9, fig.height=4.5}
i <- 2
res <- res_list[[i]]
plot_data <- res %>%
    select(eqtl_gene_name, id, lead_eqtl_pval_larger_window) %>%
    mutate(pval = -log10(lead_eqtl_pval_larger_window )) %>%
    pivot_wider(-lead_eqtl_pval_larger_window, names_from = id, values_from = pval) %>%
    tibble::column_to_rownames(var = "eqtl_gene_name") %>% 
    as.matrix()

pheatmap(mat = plot_data[c("LIMD1", "SACM1L", "SLC6A20", "LZTFL1", "CCR9", "FYCO1", "CXCR6", "XCR1"), ],
         scale = "none",
         color = viridis::viridis(100),
         na_col = "grey95",
         cluster_rows = F, cluster_cols = T)

#ggplot(data = res, aes(x = eqtl_gene_name, y = eqtl_data, fill = -log10(lead_eqtl_pval_larger_window))) + 
#  geom_tile()

```

__2) Heatmap of eQTL P-value of the GWAS variant__

* eQTL Catalogue

```{r fig-gwas-hit-eqtl-pval-heatmap-eqtl-catalogue, fig.width=12, fig.height=5}
i <- 1
res <- res_list[[i]]
plot_data <- res %>%
  select(eqtl_gene_name, id, lead_gwas_variant_eqtl_pval) %>%
  mutate(pval = -log10(lead_gwas_variant_eqtl_pval)) %>%
  pivot_wider(-lead_gwas_variant_eqtl_pval, names_from = id, values_from = pval) %>%
  tibble::column_to_rownames(var = "eqtl_gene_name") %>% 
  as.matrix()

pheatmap(mat = plot_data[c("LIMD1", "SACM1L", "SLC6A20", "LZTFL1", "CCR9", "FYCO1", "CXCR6", "XCR1"), ],
         scale = "none",
         color = viridis::viridis(100),
         na_col = "grey95",
         cluster_rows = F, cluster_cols = T)

#ggplot(data = res, aes(x = eqtl_gene_name, y = eqtl_data, fill = -log10(lead_gwas_variant_eqtl_pval))) + 
#  geom_tile()

```

* GTEx

```{r fig-gwas-hit-eqtl-pval-heatmap-gtex, fig.width=9, fig.height=4.5}
i <- 2
res <- res_list[[i]]
plot_data <- res %>%
  select(eqtl_gene_name, id, lead_gwas_variant_eqtl_pval) %>%
  mutate(pval = -log10(lead_gwas_variant_eqtl_pval)) %>%
  pivot_wider(-lead_gwas_variant_eqtl_pval, names_from = id, values_from = pval) %>%
  tibble::column_to_rownames(var = "eqtl_gene_name") %>% 
  as.matrix()

pheatmap(mat = plot_data[c("LIMD1", "SACM1L", "SLC6A20", "LZTFL1", "CCR9", "FYCO1", "CXCR6", "XCR1"), ],
         scale = "none",
         color = viridis::viridis(100),
         na_col = "grey95",
         cluster_rows = F, cluster_cols = T)

#ggplot(data = res, aes(x = eqtl_gene_name, y = eqtl_data, fill = -log10(lead_gwas_variant_eqtl_pval))) + 
#  geom_tile()

```

#### Summary of coloc-standard

* __eQTL Catalogue__

```{r fig-summary-coloc-eqtl-catalogue, fig.width=17, fig.height=14}
i <- 1
res <- res_list[[i]]

res %>%
  filter(!is.na(coloc_pp4)) %>% 
  pivot_longer(cols = starts_with("coloc_p"), names_to = "coloc_hyp", values_to = "coloc_pp") %>% 
  mutate(coloc_hyp = sub("coloc_pp", "H", coloc_hyp)) %>% 
  mutate(study = sub("_ge_", "\n", eqtl_data)) %>% 
  mutate(study = sub("_microarray_", "\n", study)) %>% 
  mutate(gene_name_f = factor(eqtl_gene_name, levels = c("LIMD1", "SACM1L", "SLC6A20", "LZTFL1", "CCR9", "FYCO1", "CXCR6", "XCR1"))) %>% 
  ggplot(aes(x = gene_name_f, y = coloc_pp, fill = coloc_hyp)) +
  labs(title = "eQTL Catalogue",
       x = "Gene name",
       y = "Posterior probability from coloc",
       fill = "Coloc hypothesis") +
  geom_col() +
  facet_wrap(~study, ncol = 7) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

```

Studies where PP4 > 0.5:

```{r}
res %>%
  filter(!is.na(coloc_pp4),
         coloc_pp4 > 0.5) %>% 
  select(-min_lead_gwas_variant_eqtl_pval,
         -id) %>% 
  mutate(lead_gwas_meta_pval = as.character(signif(lead_gwas_meta_pval, 4)),
         lead_eqtl_pval_larger_window = as.character(signif(lead_eqtl_pval_larger_window, 4)),
         lead_eqtl_pval = as.character(signif(lead_eqtl_pval, 4)),
         lead_gwas_variant_eqtl_pval = as.character(signif(lead_gwas_variant_eqtl_pval, 4))) %>% 
  kable() %>% 
  kable_styling(bootstrap_options = c("striped", "condensed"), full_width = F, font_size = 14)
  
```

* __GTEx__

```{r fig-summary-coloc-gtex, fig.width=17, fig.height=12}
i <- 2
res <- res_list[[i]]

res %>%
  filter(!is.na(coloc_pp4)) %>% 
  pivot_longer(cols = starts_with("coloc_p"), names_to = "coloc_hyp", values_to = "coloc_pp") %>% 
  mutate(coloc_hyp = sub("coloc_pp", "H", coloc_hyp)) %>% 
  mutate(study = sub("_ge_", "\n", eqtl_data)) %>% 
  mutate(study = sub("_microarray_", "\n", study)) %>% 
  mutate(gene_name_f = factor(eqtl_gene_name, levels = c("LIMD1", "SACM1L", "SLC6A20", "LZTFL1", "CCR9", "FYCO1", "CXCR6", "XCR1"))) %>% 
  ggplot(aes(x = gene_name_f, y = coloc_pp, fill = coloc_hyp)) +
  labs(title = "GTEx_v8",
       x = "Gene name",
       y = "Posterior probability from coloc",
       fill = "Coloc hypothesis") +
  geom_col() +
  facet_wrap(~study, ncol = 7) +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1))

```

Studies where PP4 > 0.5:

```{r}
res %>%
  filter(!is.na(coloc_pp4),
         coloc_pp4 > 0.5) %>% 
  select(-id) %>% 
  mutate(lead_gwas_meta_pval = as.character(signif(lead_gwas_meta_pval, 4)),
         lead_eqtl_pval_larger_window = as.character(signif(lead_eqtl_pval_larger_window, 4)),
         lead_eqtl_pval = as.character(signif(lead_eqtl_pval, 4)),
         lead_gwas_variant_eqtl_pval = as.character(signif(lead_gwas_variant_eqtl_pval, 4))) %>% 
  kable() %>% 
  kable_styling(bootstrap_options = c("striped", "condensed"), full_width = F, font_size = 14)
  
```

#### Summary coloc-condmask

__Compare coloc-standard and coloc-condmask results__

```{r fig-compare-coloc, fig.width=4.5, fig.height=4.5}
# eQTL Catalogue -------------------
stand <- res_list[[1]]
stand <- stand[!is.na(stand$coloc_pp4),]
stopifnot(length(unique(stand$id_match)) == nrow(stand))

condmask <- res_condmask_list[[1]] %>% 
  group_by(eqtl_data, eqtl_gene_name) %>% 
  arrange(desc(PP.H4.abf), .by_group = TRUE) %>% 
  ungroup() %>% 
  distinct(eqtl_gene_name, eqtl_data, .keep_all = TRUE) %>% 
  data.frame()
stopifnot(length(unique(condmask$id_match)) == nrow(condmask))

id <- intersect(stand$id_match, condmask$id_match)

plot(stand[match(id, stand$id_match), "coloc_pp4"], condmask[match(id, condmask$id_match), "PP.H4.abf"],
     xlab = "PP4 from coloc-standard", ylab = "Max PP4 from coloc-condmask",
     xlim = c(0, 1), ylim = c(0, 1),
     main = "eQTL catalogue")
abline(0, 1, col = "royalblue")

# GTEx v8 -------------------
stand <- res_list[[2]]
stand <- stand[!is.na(stand$coloc_pp4),]
stopifnot(length(unique(stand$id_match)) == nrow(stand))

condmask <- res_condmask_list[[2]] %>% 
  group_by(eqtl_data, eqtl_gene_name) %>% 
  arrange(desc(PP.H4.abf), .by_group = TRUE) %>% 
  ungroup() %>% 
  distinct(eqtl_gene_name, eqtl_data, .keep_all = TRUE) %>% 
  data.frame()
stopifnot(length(unique(condmask$id_match)) == nrow(condmask))

id <- intersect(stand$id_match, condmask$id_match)

plot(stand[match(id, stand$id_match), "coloc_pp4"], condmask[match(id, condmask$id_match), "PP.H4.abf"], 
     xlab = "PP4 from coloc-standard", ylab = "Max PP4 from coloc-condmask",
     xlim = c(0, 1), ylim = c(0, 1),
     main = "GTEx")
abline(0, 1, col = "royalblue")

```

Very similar results: no secondary independent signals seem to be colocalized with the GWAS.

```{r}
# Supplementary table 
out <- list()

# Combine coloc-standard and coloc-cond/mask results - eQTL Catalogue
stand <- res_list[[1]]
stand$lead_eqtl_variant_larger_window <- NULL
stand$lead_eqtl_pval_larger_window <- NULL
stand$min_lead_gwas_variant_eqtl_pval <- NULL
condmask <- res_condmask_list[[1]]
condmask <- condmask[!duplicated(condmask$id_match),]
stand$max_pp4_condmask <- condmask[match(stand$id_match, condmask$id_match), "max_pp4"]
stand$id <- NULL
stand$id_match <- NULL
# there's four cases where the best probe (= min lead_gwas_eqtl_pval) was not used in coloc test
# but another probe was used in coloc test, because there was a snp within 100kb from the lead gwas variant with < 10-4, but the lead gwas had eqtl p even lower => no coloc anyway
stand[is.na(stand$coloc_pp4), "max_pp4_condmask"] <- NA
out[[1]] <- stand

# Combine coloc-standard and coloc-cond/mask results - GTEx
stand <- res_list[[2]]
stand$lead_eqtl_variant_larger_window <- NULL
stand$lead_eqtl_pval_larger_window <- NULL
condmask <- res_condmask_list[[2]]
condmask <- condmask[!duplicated(condmask$id_match),]
stand$max_pp4_condmask <- condmask[match(stand$id_match, condmask$id_match), "max_pp4"]
stand$id <- NULL
stand$id_match <- NULL
out[[2]] <- stand

# write out
out <- do.call(rbind, out)
write.table(out, file = here("covid_gwas", "freeze5", "chr3_coloc", "result", paste0("suppl_table_coloc_results.", gwas_analysis, ".txt")), col.names = T, row.names = F, sep = "\t", quote = F)
```


* __eQTL Catalogue__

Studies where PP4 > 0.5:

```{r}
i <- 1
res <- res_condmask_list[[i]]

res %>%
  filter(max_pp4 > 0.5) %>% 
  select(-id,
         -id_match) %>% 
  select(eqtl_data, eqtl_gene_name, max_pp4, colnames(res)[!colnames(res) %in% c("eqtl_data", "eqtl_gene_name", "max_pp4", "id", "id_match")]) %>% 
  mutate(lead_gwas_meta_pval = as.character(signif(lead_gwas_meta_pval, 4)),
         lead_eqtl_pval_larger_window = as.character(signif(lead_eqtl_pval_larger_window, 4)),
         lead_eqtl_pval = as.character(signif(lead_eqtl_pval, 4)),
         lead_gwas_variant_eqtl_pval = as.character(signif(lead_gwas_variant_eqtl_pval, 4))) %>% 
  kable() %>% 
  kable_styling(bootstrap_options = c("striped", "condensed"), full_width = F, font_size = 14)
  
```

* __GTEx__

Studies where PP4 > 0.5:

```{r}
i <- 2
res <- res_condmask_list[[i]]

res %>%
  filter(max_pp4 > 0.5) %>% 
  select(-id,
         -id_match) %>% 
  select(eqtl_data, eqtl_gene_name, max_pp4, colnames(res)[!colnames(res) %in% c("eqtl_data", "eqtl_gene_name", "max_pp4", "id", "id_match")]) %>% 
  mutate(lead_gwas_meta_pval = as.character(signif(lead_gwas_meta_pval, 4)),
         lead_eqtl_pval_larger_window = as.character(signif(lead_eqtl_pval_larger_window, 4)),
         lead_eqtl_pval = as.character(signif(lead_eqtl_pval, 4)),
         lead_gwas_variant_eqtl_pval = as.character(signif(lead_gwas_variant_eqtl_pval, 4))) %>% 
  kable() %>% 
  kable_styling(bootstrap_options = c("striped", "condensed"), full_width = F, font_size = 14)
  
```
